### ISSUE TITLE
Convert steps into gold per hour

### ISSUE THREAD

**Issue by Iiridayn** [t-mw/gw2-arbitrage#12]
Splitting this off from #7 as it's not related.

> Recipes that require assembly could be another opt-in. Ideally I'd want to be able to keep them in the default output, and accurately calculate the time required for each recipe. Then recipes could be ranked by gold per time. Unfortunately as far as I could tell the crafting times from the API weren't accurate, so I added crafting steps as an approximation of time required.

I noted:

I've also been thinking about a gold per hour calculation, and recognized steps as a proxy for it; I hadn't realized the crafting times were included in the API. Could it be the API crafting time is for a single item? Crafting multiple items provides a speed bonus for each - https://wiki.guildwars2.com/wiki/Crafting says:

When crafting a stack of items, the process speeds up progressively: each item queued will craft in half the time of the previous item. For example, if it takes two seconds to create a single bronze ingot, it will take only one second to create the second, and a half second to craft a third. There is a limit to how fast items can be crafted.

I suspect the halving ends after 4-5 times; but it should be pretty easy to craft a stack of mithril ore or something (do it all the time...) and time the duration. I'll check it out.

Two problems with assembler recipes:

1. Users might not have access to the assembler in any guild, or the guild may not have the recipe unlocked
2. Crafting assembler recipes ties you to the computer every 5.5 minutes, unlike other recipes which you autocraft many stacks of. Showing an accurate crafting duration might reasonably answer this point though. Also, "many stacks" still is almost universally less than 5.5 minutes.

**Comment by t-mw** [comment]
> baseTime / 1.572 ^ floor(index, 5) (index is 0 based) per item is almost correct

Thanks for looking into this. Did you settle on 1.572 through trial and error? Even if it's not perfect it sounds accurate enough to be more useful than what the tool currently shows.

**Comment by Iiridayn** [comment]
Yes, trial and error. I guessed (as does the wiki) this was an exponential function of the form C^n (C being a constant). The slope possibly doesn't _quite_ match, but that is the most common kind of function I've seen in games (also used for XP curves). The constant was determined with a manual binary search; not aiming for terribly high accuracy, so it could be refined. One thing that makes me uncomfortable about it is I can't find a good ratio or well known irrational number which fits it. People _usually_ prefer familiar numbers, numbers with 5's and 0's, and numbers with few decimal places. It is possible the number was picked via trial and error as well - but I feel that is improbable. The curve not quite matching also confirms I'm missing something important somewhere.

That all said - yeah, it seems close enough to be _useful_ - which maybe is all one can ask of it without knowing the proper formula :).

**Comment by Iiridayn** [comment]
Another stopwatch value. 1500 items at 1000 ms each took 244.04 seconds. I add this since the average time should be quite close to the minimum time, and I was crafting them anyway.

Notably, this is < 1/6 * base time per item, but only just.

**Comment by thangtq-pionero** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
vcccaat, @hakeemullahjan, @SonnyCastro, @Lpfigueiredo, @bn-l, @desertainment, @michabre, @RustMan88, @Azurine, @xuhcc, @DonnyZhao, @nichosystem, @boombb12138, @wukong010, @cademccu, @vamsiikrishna, @uurtech, @sasharussia43, @ichelsearay, @Oscillator-Phoenix
